siz2fastq.c

```diff
@@ -1,96 +1,217 @@
+#include <stdbool.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include <zstd.h>
 #include <zlib.h>
+#include <unistd.h>
 
 #define BUFFER_SIZE 4096
-#define LINE_BUFFER_SIZE 1024 * 1024
 
-int main(int argc, char *argv[]) {
-  if (argc != 3) {
+// Split an interleaved zstd-compressed fastq file into a pair of optionally
+// gzip-compressed fastq files.  No input validation is performed: the input
+// must be valid fastq with no hard wrapping or you'll get garbage output.
+
+// Struct to track what kind of file we're writing, since we optionally
+// compress our output.
+typedef struct {
+    FILE* regular_file;
+    gzFile gz_file;
+    bool is_gzipped;
+} OutputFile;
+
+OutputFile open_output_file(const char* filename, bool use_gzip) {
+    OutputFile output;
+    output.is_gzipped = use_gzip;
+
+    if (use_gzip) {
+        output.gz_file = gzopen(filename, "wb");
+        if (output.gz_file == Z_NULL) {
+            fprintf(stderr,
+                    "Unable to open %s as a gzip-compressed output file\n",
+                    filename);
+            exit(1);
+        }
+        output.regular_file = NULL;
+    } else {
+        output.regular_file = fopen(filename, "wb");
+        if (output.regular_file == NULL) {
+            fprintf(stderr, "Unable to open %s as output file\n", filename);
+            exit(1);
+        }
+        output.gz_file = NULL;
+    }
+
+    return output;
+}
+
+// Write `size` bytes from `buffer` to `file` or exit with an error message.
+void write_output_or_die(
+         OutputFile* file, const void* buffer, size_t size) {
+    size_t bytes_written;
+    if (file->is_gzipped) {
+        bytes_written = gzwrite(file->gz_file, buffer, size);
+    } else {
+        bytes_written = fwrite(buffer, /*nitems=*/1, size, file->regular_file);
```

Comment:
Isn't the second argument to `fwrite` the size of each item, not the number of items?

Reply:
(Also applies on line 140)

Reply:
Good point!  My comment was confusing and wrong.  Edited!